File: backend/src/services/feedbackService.js
// src/services/feedbackService.js
// Avaliações da homepage corporativa (secção "Avaliações Verificadas").
//
// Só quem tem sessão e empresa na KIXIMA pode avaliar — nunca anónimo. É
// exatamente isso que torna o selo "Verificado" real: a autoria vem sempre de
// req.user (nome/empresa nunca são digitados à mão), e o alvo da avaliação
// (fornecedor, produto, pedido, entrega, pagamento, atendimento) tem de ser
// algo que a empresa do autor realmente viveu — validado em resolverAlvo()
// antes de guardar. "Experiência geral" é a única categoria sem alvo.
//
// Uma avaliação só aparece na home depois de aprovada pelo Admin do Sistema
// (Suporte); a média mostrada conta sempre TODAS as aprovadas, não só as
// exibidas na parede.
const prisma = require('../config/database');
const { NotFoundError, ValidationError } = require('../utils/errors');
const paginacao = require('../utils/paginacao');

const CATEGORIAS = ['FORNECEDOR', 'PRODUTO', 'SERVICO', 'PEDIDO', 'ENTREGA', 'PAGAMENTO', 'ATENDIMENTO', 'EXPERIENCIA_GERAL'];
const MENSAGEM_MAX = 700;
const OPCOES_LIMITE = 30; // por categoria — o suficiente para escolher sem paginar um dropdown

function limpar(valor, max) {
  return typeof valor === 'string' ? valor.trim().slice(0, max) : '';
}

// Usado por minhas() e pela moderação do Admin do Sistema — inclui
// `approved`, para quem tem de saber em que estado a avaliação está.
const SUMMARY = { id: true, user: { select: { name: true } }, company: { select: { name: true } }, categoria: true, targetLabel: true, rating: true, message: true, verified: true, approved: true, createdAt: true };
// A parede pública nunca mostra `approved` — é um detalhe interno de
// moderação, não algo para expor a quem só lê a homepage.
const PUBLIC_SUMMARY = { id: true, user: { select: { name: true } }, company: { select: { name: true } }, categoria: true, targetLabel: true, rating: true, message: true, verified: true, createdAt: true };

/**
 * Opções reais para o dropdown de "sobre o que é esta avaliação", por
 * categoria — construídas a partir do histórico real da empresa do
 * utilizador (como compradora OU fornecedora, conforme o lado em que
 * participou), nunca inventadas. Ordenadas por mais recente.
 */
async function opcoes({ companyId, userId }) {
  if (!companyId) return { FORNECEDOR: [], PRODUTO: [], SERVICO: [], PEDIDO: [], ENTREGA: [], PAGAMENTO: [], ATENDIMENTO: [] };

  const [pedidos, tickets] = await Promise.all([
    prisma.purchaseOrder.findMany({
      where: { OR: [{ buyerCompanyId: companyId }, { supplierCompanyId: companyId }] },
      orderBy: { createdAt: 'desc' },
      take: 100,
      select: {
        id: true, reference: true, createdAt: true, deliveredAt: true, receivedAt: true,
        buyerCompanyId: true, supplierCompanyId: true,
        buyerCompany: { select: { id: true, name: true } },
        supplierCompany: { select: { id: true, name: true } },
        items: { select: { product: { select: { id: true, name: true, kind: true } } } },
        invoice: { select: { payment: { select: { id: true, reference: true, status: true, processedAt: true } } } },
      },
    }),
    prisma.supportTicket.findMany({
      where: { OR: [{ companyId }, { userId }] },
      orderBy: { createdAt: 'desc' },
      take: OPCOES_LIMITE,
      select: { id: true, reference: true, subject: true, createdAt: true },
    }),
  ]);

  const fornecedores = new Map();
  const produtos = new Map();
  const servicos = new Map();
  const entregas = [];
  const pagamentos = [];

  const pedidosOpts = pedidos.map((po) => {
    const souComprador = po.buyerCompanyId === companyId;
    const contraparte = souComprador ? po.supplierCompany : po.buyerCompany;
    if (contraparte && !fornecedores.has(contraparte.id)) {
      fornecedores.set(contraparte.id, { id: contraparte.id, label: contraparte.name });
    }
    for (const item of po.items) {
      const p = item.product;
      if (!p) continue;
      const alvo = p.kind === 'SERVICO' ? servicos : produtos;
      if (!alvo.has(p.id)) alvo.set(p.id, { id: p.id, label: p.name });
    }
    if (po.deliveredAt || po.receivedAt) {
      entregas.push({ id: po.id, label: `${po.reference} — ${contraparte?.name || ''}` });
    }
    const pagamento = po.invoice?.payment;
    if (pagamento && pagamento.status === 'PROCESSADO') {
      pagamentos.push({ id: pagamento.id, label: `${pagamento.reference} — ${po.reference}` });
    }
    return { id: po.id, label: `${po.reference} — ${contraparte?.name || ''}` };
  });

  return {
    FORNECEDOR: [...fornecedores.values()].slice(0, OPCOES_LIMITE),
    PRODUTO: [...produtos.values()].slice(0, OPCOES_LIMITE),
    SERVICO: [...servicos.values()].slice(0, OPCOES_LIMITE),
    PEDIDO: pedidosOpts.slice(0, OPCOES_LIMITE),
    ENTREGA: entregas.slice(0, OPCOES_LIMITE),
    PAGAMENTO: pagamentos.slice(0, OPCOES_LIMITE),
    ATENDIMENTO: tickets.map((t) => ({ id: t.id, label: `${t.reference} — ${t.subject}` })),
  };
}

/**
 * Confirma que o alvo escolhido pertence mesmo ao histórico da empresa do
 * autor (nunca confiar num id vindo do cliente) e devolve o rótulo a
 * guardar como snapshot. Atira ValidationError se não encontrar nada —
 * silenciosamente aceitar um id qualquer tornaria o selo "Verificado" uma
 * mentira.
 */
async function resolverAlvo({ categoria, targetId, companyId, userId }) {
  if (categoria === 'EXPERIENCIA_GERAL') return { targetId: null, targetLabel: null };

  if (!targetId) throw new ValidationError('Escolha a que se refere esta avaliação.');

  const naoEncontrado = () => new ValidationError('Não encontrámos esse registo no histórico da sua empresa.');

  if (categoria === 'FORNECEDOR') {
    const po = await prisma.purchaseOrder.findFirst({
      where: {
        OR: [
          { buyerCompanyId: companyId, supplierCompanyId: targetId },
          { supplierCompanyId: companyId, buyerCompanyId: targetId },
        ],
      },
      select: { buyerCompany: { select: { id: true, name: true } }, supplierCompany: { select: { id: true, name: true } }, buyerCompanyId: true },
    });
    if (!po) throw naoEncontrado();
    const contraparte = po.buyerCompanyId === companyId ? po.supplierCompany : po.buyerCompany;
    return { targetId, targetLabel: contraparte.name };
  }

  if (categoria === 'PRODUTO' || categoria === 'SERVICO') {
    const item = await prisma.purchaseOrderItem.findFirst({
      where: {
        productId: targetId,
        purchaseOrder: { OR: [{ buyerCompanyId: companyId }, { supplierCompanyId: companyId }] },
      },
      select: { product: { select: { name: true, kind: true } } },
    });
    if (!item || item.product.kind !== categoria) throw naoEncontrado();
    return { targetId, targetLabel: item.product.name };
  }

  if (categoria === 'PEDIDO' || categoria === 'ENTREGA') {
    const po = await prisma.purchaseOrder.findFirst({
      where: {
        id: targetId,
        OR: [{ buyerCompanyId: companyId }, { supplierCompanyId: companyId }],
      },
      select: { reference: true, buyerCompanyId: true, buyerCompany: { select: { name: true } }, supplierCompany: { select: { name: true } }, deliveredAt: true, receivedAt: true },
    });
    if (!po) throw naoEncontrado();
    if (categoria === 'ENTREGA' && !po.deliveredAt && !po.receivedAt) throw naoEncontrado();
    const contraparte = po.buyerCompanyId === companyId ? po.supplierCompany : po.buyerCompany;
    return { targetId, targetLabel: `${po.reference} — ${contraparte?.name || ''}` };
  }

  if (categoria === 'PAGAMENTO') {
    const pagamento = await prisma.payment.findFirst({
      where: {
        id: targetId,
        status: 'PROCESSADO',
        invoice: { purchaseOrder: { OR: [{ buyerCompanyId: companyId }, { supplierCompanyId: companyId }] } },
      },
      select: { reference: true, invoice: { select: { purchaseOrder: { select: { reference: true } } } } },
    });
    if (!pagamento) throw naoEncontrado();
    return { targetId, targetLabel: `${pagamento.reference} — ${pagamento.invoice?.purchaseOrder?.reference || ''}` };
  }

  if (categoria === 'ATENDIMENTO') {
    const ticket = await prisma.supportTicket.findFirst({
      where: { id: targetId, OR: [{ companyId }, { userId }] },
      select: { reference: true, subject: true },
    });
    if (!ticket) throw naoEncontrado();
    return { targetId, targetLabel: `${ticket.reference} — ${ticket.subject}` };
  }

  throw naoEncontrado();
}

/** Submissão pelo utilizador autenticado — nome e empresa vêm sempre da sessão. */
async function criar({ userId, companyId, categoria, targetId, rating, message }) {
  if (!companyId) throw new ValidationError('É necessário pertencer a uma empresa para enviar uma avaliação.');
  if (!CATEGORIAS.includes(categoria)) throw new ValidationError(`Categoria inválida. Escolha uma de: ${CATEGORIAS.join(', ')}.`);

  const ratingNum = Number(rating);
  if (!Number.isInteger(ratingNum) || ratingNum < 1 || ratingNum > 5) {
    throw new ValidationError('Selecione uma classificação entre 1 e 5.');
  }

  const mensagemLimpa = limpar(message, MENSAGEM_MAX);
  if (!mensagemLimpa) throw new ValidationError('Escreva um comentário.');

  const alvo = await resolverAlvo({ categoria, targetId, companyId, userId });

  const criado = await prisma.feedback.create({
    data: {
      userId, companyId, categoria,
      targetId: alvo.targetId, targetLabel: alvo.targetLabel,
      rating: ratingNum, message: mensagemLimpa,
      verified: true,
    },
  });
  return { recebido: true, id: criado.id };
}

/** As avaliações do próprio utilizador (qualquer estado) — para acompanhar o que já enviou. */
async function minhas({ userId }) {
  return prisma.feedback.findMany({
    where: { userId },
    orderBy: { createdAt: 'desc' },
    select: SUMMARY,
  });
}

/**
 * O que a homepage pública mostra: as avaliações aprovadas mais recentes
 * (limitadas, para a parede não crescer sem fim) e a média de TODAS as
 * aprovadas — não só as exibidas, para o número não mentir por omissão.
 */
async function publicar() {
  const where = { approved: true };
  const [recentes, agregados] = await Promise.all([
    prisma.feedback.findMany({
      where,
      orderBy: { createdAt: 'desc' },
      take: 6,
      select: PUBLIC_SUMMARY,
    }),
    prisma.feedback.aggregate({ where, _avg: { rating: true }, _count: true }),
  ]);

  return {
    feedback: recentes,
    total: agregados._count,
    average: agregados._count ? Number(agregados._avg.rating.toFixed(1)) : 0,
  };
}

/** Fila de moderação do Admin do Sistema — tudo, aprovado ou não, mais recente primeiro. */
async function listarAdmin({ page, limit, status } = {}) {
  const p = paginacao.parametros({ page, limit });
  const where = status === 'pendente' ? { approved: false } : status === 'aprovado' ? { approved: true } : {};

  const [total, itens] = await Promise.all([
    prisma.feedback.count({ where }),
    prisma.feedback.findMany({ where, orderBy: { createdAt: 'desc' }, skip: p.skip, take: p.take, select: SUMMARY }),
  ]);
  return paginacao.envelope(itens, total, p);
}

async function aprovar(id) {
  const existe = await prisma.feedback.findUnique({ where: { id } });
  if (!existe) throw new NotFoundError('Avaliação');
  return prisma.feedback.update({ where: { id }, data: { approved: true }, select: SUMMARY });
}

/** Rejeitar É remover — não há um terceiro estado "rejeitado" a mostrar a ninguém. */
async function remover(id) {
  const existe = await prisma.feedback.findUnique({ where: { id } });
  if (!existe) throw new NotFoundError('Avaliação');
  await prisma.feedback.delete({ where: { id } });
}

module.exports = { CATEGORIAS, opcoes, criar, minhas, publicar, listarAdmin, aprovar, remover };

File: backend-java/src/main/java/ao/kixima/feedback/FeedbackService.java
package ao.kixima.feedback;

import ao.kixima.common.error.NotFoundException;
import ao.kixima.common.error.ValidationException;
import ao.kixima.common.pagination.PaginaResposta;
import ao.kixima.common.pagination.Paginacao;
import ao.kixima.company.Company;
import ao.kixima.feedback.dto.CreateFeedbackRequest;
import ao.kixima.feedback.dto.FeedbackCreatedDto;
import ao.kixima.feedback.dto.FeedbackDto;
import ao.kixima.feedback.dto.FeedbackNameRef;
import ao.kixima.feedback.dto.FeedbackOptionDto;
import ao.kixima.feedback.dto.FeedbackOptionsResponse;
import ao.kixima.feedback.dto.PublicFeedbackResponse;
import ao.kixima.catalog.ProductKind;
import ao.kixima.payment.Payment;
import ao.kixima.payment.PaymentRepository;
import ao.kixima.payment.PaymentStatus;
import ao.kixima.po.PurchaseOrder;
import ao.kixima.po.PurchaseOrderItem;
import ao.kixima.po.PurchaseOrderItemRepository;
import ao.kixima.po.PurchaseOrderRepository;
import ao.kixima.support.SupportTicket;
import ao.kixima.support.SupportTicketRepository;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.Instant;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.UUID;

/**
 * Espelha backend/src/services/feedbackService.js — avaliações da homepage
 * corporativa ("Avaliações Verificadas"). Só quem tem sessão e empresa na
 * KIXIMA avalia; o alvo (fornecedor, produto, pedido, entrega, atendimento)
 * tem de ser algo que a empresa do autor realmente viveu (ver
 * {@link #resolverAlvo}) — nunca aceite de olhos fechados a partir do
 * cliente. "Experiência geral" é a única categoria sem alvo.
 *
 * A categoria PAGAMENTO verifica-se contra {@link Payment} (PROCESSADO, de
 * uma PO em que a empresa é parte) — portado com o domínio de pagamentos
 * (grupo A das lacunas pós-M6).
 */
@Service
public class FeedbackService {

    private static final int MAX_OPCOES = 30;
    private static final int MAX_MENSAGEM = 700;

    private final FeedbackRepository feedbackRepository;
    private final PurchaseOrderRepository purchaseOrderRepository;
    private final PurchaseOrderItemRepository purchaseOrderItemRepository;
    private final SupportTicketRepository supportTicketRepository;
    private final PaymentRepository paymentRepository;

    public FeedbackService(FeedbackRepository feedbackRepository, PurchaseOrderRepository purchaseOrderRepository,
                            PurchaseOrderItemRepository purchaseOrderItemRepository, SupportTicketRepository supportTicketRepository,
                            PaymentRepository paymentRepository) {
        this.paymentRepository = paymentRepository;
        this.feedbackRepository = feedbackRepository;
        this.purchaseOrderRepository = purchaseOrderRepository;
        this.purchaseOrderItemRepository = purchaseOrderItemRepository;
        this.supportTicketRepository = supportTicketRepository;
    }

    /** Opções reais para o dropdown "sobre o que é esta avaliação" — construídas do histórico real da empresa. */
    @Transactional(readOnly = true)
    public FeedbackOptionsResponse opcoes(String companyId, String userId) {
        if (companyId == null) {
            return new FeedbackOptionsResponse(List.of(), List.of(), List.of(), List.of(), List.of(), List.of(), List.of());
        }
        List<PurchaseOrder> pedidos = purchaseOrderRepository.findByCompanyIdOrderByCreatedAtDesc(companyId, Pageable.ofSize(100));
        List<SupportTicket> tickets = supportTicketRepository.findByCompanyIdOrUserIdOrderByCreatedAtDesc(companyId, userId, Pageable.ofSize(MAX_OPCOES));

        Map<String, FeedbackOptionDto> fornecedores = new LinkedHashMap<>();
        Map<String, FeedbackOptionDto> produtos = new LinkedHashMap<>();
        Map<String, FeedbackOptionDto> servicos = new LinkedHashMap<>();
        List<FeedbackOptionDto> pedidosOpts = new java.util.ArrayList<>();
        List<FeedbackOptionDto> entregas = new java.util.ArrayList<>();
        List<FeedbackOptionDto> pagamentos = new java.util.ArrayList<>();

        // `po.invoice.payment` do Node — um SELECT só para os pagamentos das 100 POs.
        Map<String, Payment> pagamentoPorFatura = new LinkedHashMap<>();
        List<String> invoiceIds = pedidos.stream().map(PurchaseOrder::getInvoice).filter(java.util.Objects::nonNull)
                .map(ao.kixima.invoice.Invoice::getId).toList();
        if (!invoiceIds.isEmpty()) {
            for (Payment p : paymentRepository.findByInvoiceIdIn(invoiceIds)) pagamentoPorFatura.put(p.getInvoiceId(), p);
        }

        for (PurchaseOrder po : pedidos) {
            boolean souComprador = po.getBuyerCompanyId().equals(companyId);
            Company contraparte = souComprador ? po.getSupplierCompany() : po.getBuyerCompany();
            if (contraparte != null) fornecedores.putIfAbsent(contraparte.getId(), new FeedbackOptionDto(contraparte.getId(), contraparte.getName()));
            for (PurchaseOrderItem item : po.getItems()) {
                var produto = item.getProduct();
                if (produto == null) continue;
                Map<String, FeedbackOptionDto> alvo = produto.getKind() == ProductKind.SERVICO ? servicos : produtos;
                alvo.putIfAbsent(produto.getId(), new FeedbackOptionDto(produto.getId(), produto.getName()));
            }
            String rotulo = po.getReference() + " — " + (contraparte != null ? contraparte.getName() : "");
            pedidosOpts.add(new FeedbackOptionDto(po.getId(), rotulo));
            if (po.getDeliveredAt() != null || po.getReceivedAt() != null) {
                entregas.add(new FeedbackOptionDto(po.getId(), rotulo));
            }
            Payment pagamento = po.getInvoice() == null ? null : pagamentoPorFatura.get(po.getInvoice().getId());
            if (pagamento != null && pagamento.getStatus() == PaymentStatus.PROCESSADO) {
                pagamentos.add(new FeedbackOptionDto(pagamento.getId(), pagamento.getReference() + " — " + po.getReference()));
            }
        }

        List<FeedbackOptionDto> atendimento = tickets.stream()
                .map(t -> new FeedbackOptionDto(t.getId(), t.getReference() + " — " + t.getSubject()))
                .toList();

        return new FeedbackOptionsResponse(
                limitar(fornecedores.values()), limitar(produtos.values()), limitar(servicos.values()),
                limitar(pedidosOpts), limitar(entregas), limitar(pagamentos), limitar(atendimento));
    }

    private List<FeedbackOptionDto> limitar(java.util.Collection<FeedbackOptionDto> valores) {
        return valores.stream().limit(MAX_OPCOES).toList();
    }

    /**
     * Confirma que o alvo escolhido pertence mesmo ao histórico da empresa
     * do autor e devolve o rótulo a guardar como snapshot — aceitar um id
     * qualquer sem verificar tornaria o selo "Verificado" uma mentira.
     */
    private String[] resolverAlvo(FeedbackCategoria categoria, String targetId, String companyId, String userId) {
        if (categoria == FeedbackCategoria.EXPERIENCIA_GERAL) return new String[]{null, null};
        if (targetId == null || targetId.isBlank()) throw new ValidationException("Escolha a que se refere esta avaliação.");

        ValidationException naoEncontrado = new ValidationException("Não encontrámos esse registo no histórico da sua empresa.");

        switch (categoria) {
            case FORNECEDOR -> {
                List<PurchaseOrder> pos = purchaseOrderRepository.findEntreEmpresas(companyId, targetId, Pageable.ofSize(1));
                if (pos.isEmpty()) throw naoEncontrado;
                PurchaseOrder po = pos.get(0);
                Company contraparte = po.getBuyerCompanyId().equals(companyId) ? po.getSupplierCompany() : po.getBuyerCompany();
                return new String[]{targetId, contraparte.getName()};
            }
            case PRODUTO, SERVICO -> {
                List<PurchaseOrderItem> itens = purchaseOrderItemRepository.findByProductIdEComEmpresa(targetId, companyId, Pageable.ofSize(1));
                if (itens.isEmpty() || itens.get(0).getProduct().getKind() != ProductKind.valueOf(categoria.name())) throw naoEncontrado;
                return new String[]{targetId, itens.get(0).getProduct().getName()};
            }
            case PEDIDO, ENTREGA -> {
                PurchaseOrder po = purchaseOrderRepository.findById(targetId).orElseThrow(() -> naoEncontrado);
                boolean pertence = po.getBuyerCompanyId().equals(companyId) || po.getSupplierCompanyId().equals(companyId);
                if (!pertence) throw naoEncontrado;
                if (categoria == FeedbackCategoria.ENTREGA && po.getDeliveredAt() == null && po.getReceivedAt() == null) throw naoEncontrado;
                Company contraparte = po.getBuyerCompanyId().equals(companyId) ? po.getSupplierCompany() : po.getBuyerCompany();
                return new String[]{targetId, po.getReference() + " — " + (contraparte != null ? contraparte.getName() : "")};
            }
            case PAGAMENTO -> {
                Payment pagamento = paymentRepository.findByIdComFatura(targetId).orElseThrow(() -> naoEncontrado);
                PurchaseOrder po = pagamento.getInvoice().getPurchaseOrder();
                boolean daEmpresa = po != null && (po.getBuyerCompanyId().equals(companyId) || po.getSupplierCompanyId().equals(companyId));
                if (pagamento.getStatus() != PaymentStatus.PROCESSADO || !daEmpresa) throw naoEncontrado;
                return new String[]{targetId, pagamento.getReference() + " — " + po.getReference()};
            }
            case ATENDIMENTO -> {
                SupportTicket ticket = supportTicketRepository.findById(targetId).orElseThrow(() -> naoEncontrado);
                boolean pertence = (ticket.getCompanyId() != null && ticket.getCompanyId().equals(companyId)) || ticket.getUserId().equals(userId);
                if (!pertence) throw naoEncontrado;
                return new String[]{targetId, ticket.getReference() + " — " + ticket.getSubject()};
            }
            default -> throw naoEncontrado;
        }
    }

    /** Submissão pelo utilizador autenticado — nome e empresa vêm sempre da sessão. */
    @Transactional
    public FeedbackCreatedDto criar(String userId, String companyId, String categoriaBruta, String targetId, Integer rating, String message) {
        if (companyId == null) throw new ValidationException("É necessário pertencer a uma empresa para enviar uma avaliação.");
        FeedbackCategoria categoria = categoriaValida(categoriaBruta);

        if (rating == null || rating < 1 || rating > 5) throw new ValidationException("Selecione uma classificação entre 1 e 5.");

        String mensagemLimpa = message == null ? "" : message.trim();
        if (mensagemLimpa.length() > MAX_MENSAGEM) mensagemLimpa = mensagemLimpa.substring(0, MAX_MENSAGEM);
        if (mensagemLimpa.isEmpty()) throw new ValidationException("Escreva um comentário.");

        String[] alvo = resolverAlvo(categoria, targetId, companyId, userId);

        Feedback criado = new Feedback(UUID.randomUUID().toString(), userId, companyId, categoria, alvo[0], alvo[1],
                rating, mensagemLimpa, Instant.now());
        feedbackRepository.save(criado);
        return new FeedbackCreatedDto(true, criado.getId());
    }

    private FeedbackCategoria categoriaValida(String categoriaBruta) {
        try {
            return FeedbackCategoria.valueOf(categoriaBruta);
        } catch (Exception e) {
            throw new ValidationException("Categoria inválida. Escolha uma de: "
                    + String.join(", ", java.util.Arrays.stream(FeedbackCategoria.values()).map(Enum::name).toList()) + ".");
        }
    }

    @Transactional(readOnly = true)
    public List<FeedbackDto> minhas(String userId) {
        return feedbackRepository.findByUserIdOrderByCreatedAtDesc(userId).stream().map(this::toDto).toList();
    }

    /** A parede pública: as aprovadas mais recentes + a média de TODAS as aprovadas. */
    @Transactional(readOnly = true)
    public PublicFeedbackResponse publicar() {
        List<FeedbackDto> recentes = feedbackRepository.findTopApprovedOrderByCreatedAtDesc(Pageable.ofSize(6)).stream()
                .map(f -> toDto(f).semEstadoDeModeracao()).toList();
        long total = feedbackRepository.countByApproved(true);
        Double media = feedbackRepository.mediaAprovadas();
        double average = total > 0 && media != null ? Math.round(media * 10) / 10.0 : 0;
        return new PublicFeedbackResponse(recentes, total, average);
    }

    /** Fila de moderação do Admin do Sistema — tudo, aprovado ou não, mais recente primeiro. */
    @Transactional(readOnly = true)
    public PaginaResposta<FeedbackDto> listarAdmin(Integer page, Integer limit, String status) {
        Pageable pageable = Paginacao.parametros(page, limit);
        Page<Feedback> pagina = "pendente".equals(status) ? feedbackRepository.findByApprovedOrderByCreatedAtDesc(false, pageable)
                : "aprovado".equals(status) ? feedbackRepository.findByApprovedOrderByCreatedAtDesc(true, pageable)
                : feedbackRepository.findAllByOrderByCreatedAtDesc(pageable);
        return Paginacao.envelope(pagina.map(this::toDto));
    }

    @Transactional
    public FeedbackDto aprovar(String id) {
        Feedback feedback = feedbackRepository.findById(id).orElseThrow(() -> new NotFoundException("Avaliação"));
        feedback.setApproved(true);
        return toDto(feedback);
    }

    @Transactional
    public void remover(String id) {
        Feedback feedback = feedbackRepository.findById(id).orElseThrow(() -> new NotFoundException("Avaliação"));
        feedbackRepository.delete(feedback);
    }

    private FeedbackDto toDto(Feedback f) {
        return new FeedbackDto(f.getId(),
                new FeedbackNameRef(f.getUser() == null ? null : f.getUser().getName()),
                new FeedbackNameRef(f.getCompany() == null ? null : f.getCompany().getName()),
                f.getCategoria().name(), f.getTargetLabel(), f.getRating(), f.getMessage(), f.isVerified(),
                f.isApproved(), f.getCreatedAt());
    }
}
